Feature/zenko 716 management process

jeremyds opened the pull request:
# Pull request template

## Description

### Motivation and context

Why is this change required? What problem does it solve?

### Related issues

Please use the following link syntaxes #600 to reference issues in the
current repository

## Checklist

### Add tests to cover the changes

New tests added or existing tests modified to cover all changes

### Code conforms with the [style guide](https://github.com/scality/Guidelines/blob/master/CONTRIBUTING.md#coding-style-guidelines)

### Sign your work

In order to contribute to the project, you must sign your work
https://github.com/scality/Guidelines/blob/master/CONTRIBUTING.md#sign-your-work

Thank you again for contributing! We will try to test and integrate the change
as soon as we can.

jeremyds committed:
ft: ZENKO-716: stop using management code in S3

jeremyds committed:
ft: ZENKO-716: add management message encoding helper function

Testing the management process feature will require to send this kind of
message.

jeremyds committed:
ft: ZENKO-716: management, use constant and env variables for tests

jeremyds committed:
ft: ZENKO-716: mninor error log improvement

Add the url and the headers in the error log used when the stat push
request from management to orbit fails.

bert-e commented:
# Hello jeremyds,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get
information on this process.





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rahulreddy

* @dora-korpar

* @RachedBenMustapha

jeremyds commented:
@bert-e help

bert-e commented:
# Help page



The following options and commands are available at this time.

## Options ##


name   | description  | privileged
------ | ------------ | ------------
:arrow_right: **bypass_author_approval** | Bypass the pull request author's approval |  :star: 
:arrow_right: **bypass_peer_approval** | Bypass the pull request peers' approval |  :star: 
:arrow_right: **wait** | Instruct Bert-E not to run until further notice. | 
:arrow_right: **no_octopus** | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead | 
:arrow_right: **unanimity** | Change review acceptance criteria from `one reviewer at least` to `all reviewers` | 
:arrow_right: **bypass_jira_check** | Bypass the Jira issue check |  :star: 
:arrow_right: **after_pull_request** | Wait for the given pull request id to be merged before continuing with the current one. | 
:arrow_right: **bypass_commit_size** | Bypass the check on the size of the changeset ```TBA``` |  :star: 
:arrow_right: **bypass_leader_approval** | Bypass the pull request leaders' approval |  :star: 
:arrow_right: **bypass_incompatible_branch** | Bypass the check on the source branch prefix |  :star: 
:arrow_right: **bypass_build_status** | Bypass the build and test status |  :star: 
:arrow_right: **create_pull_requests** | Allow the creation of integration pull requests. | 



## Commands ##


name   | description  | privileged
------ | ------------ | ------------
:arrow_right: **retry** | Re-start a fresh build ```TBA``` | 
:arrow_right: **status** | Print Bert-E's current status in the pull request ```TBA``` | 
:arrow_right: **force_reset** | Delete integration branches & pull requests, and restart merge process from the beginning. | 
:arrow_right: **clear** | Remove all comments from Bert-E from the history ```TBA``` | 
:arrow_right: **reset** | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
:arrow_right: **help** | Print Bert-E's manual in the pull request. | 
:arrow_right: **build** | Re-start a fresh build ```TBA``` |

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rahulreddy

* @dora-korpar

* @RachedBenMustapha

jeremyds commented:
@bert-e after_pull_request=1542

bert-e commented:
# Waiting for other pull request(s)



The current pull request is locked by the **after_pull_request** option.

In order for me to merge this pull request, run the following actions first:


:arrow_right: Merge the `OPEN` pull request:

* pull request #1542





Alternatively, delete all the **after_pull_request** comments from this pull request.





*The following options are set:* **after_pull_request**

jeremyds commented on tests/functional/managementAgent/managementProcess.js:
```diff
@@ -0,0 +1,368 @@
+                    response.end();
+                    return;
+                }
+                response.writeHead(200, { 'Content-type': 'text/plan' });
```
plain

jeremyds commented on tests/functional/managementAgent/managementProcess.js:
```diff
@@ -0,0 +1,368 @@
+
+                    /* Check we receive the second overlay. */
+                    assert.strictEqual(msg.payload.toString(), body);
+
```
ws.terminate

jeremyds commented on tests/functional/managementAgent/managementProcess.js:
```diff
@@ -0,0 +1,368 @@
+           });
+       });
+
+    describe('with orbit simulator', () => {
```
not usefull anymore to have a describe specific to tests with orbit. It was done because I wanted to use before/after Each hooks but finally did not, so let's get an indentation level back.

jeremyds commented:
I deactivated file eve stage to test mine, I am facing ressource issue (CPU and then RAM) and before to handle this problem I want to make my stage and test suite run.
tl;drL DO NOT MERGE

rachedbenmustapha commented on lib/management/push.js:
```diff
@@ -92,7 +89,9 @@ function startWSManagementClient(pushEndpoint, instanceId, token) {
     const agent = createWSAgent(pushEndpoint, process.env, logger);
 
-    const url = `${pushEndpoint}/${instanceId}/ws`;
+    const url = process.env.CI_ORBIT_PUSH_ENDPOINT ||
```
Since the push endpoint is already configurable, this could follow the same pattern as for the API endpoint, instead of having a new CI-related env var.

jeremyds replied:
Which endpoint are you speaking about? The one used here (line 93) is not configurable, it is hardocded to 'https://push.api.zenko.io/api/v1/instance' in lib/management/index.js
That being said I don't know if there is other use case than tests to change that, the PUSH endpoint is not supposed to change, is it?

rachedbenmustapha replied:
There is a separate `PUSH_ENDPOINT` env var that is used and is configurable via the helm chart values.

It is different for dev, staging, production and laptop. It's even planned to have one per-customer for specific needs.

rachedbenmustapha commented on lib/management/constants.js:
```diff
@@ -0,0 +1,15 @@
+/* From RFC6455 which defines some reserved status code ranges, the range
+ * 4000-4999 are for private use. */
+const WS_STATUS_IDDLE = {
```
Typo: IDLE?

jeremyds replied:
done

rachedbenmustapha commented on package.json:
```diff
@@ -89,12 +89,13 @@
     "start_dataserver": "node dataserver.js",
     "start_s3server": "node index.js",
     "start_dmd": "npm-run-all --parallel start_mdserver start_dataserver",
+    "start_management": "node managementAgent.js",
```
@ssalaues when this PR is merged (and a new docker image built) the cloudserver-manager pod will need to use this command line

rachedbenmustapha reviewed (changes requested):
One general comment that applies to the new env vars: they are all legitimate tunables so I don't think they need to have the `CI_` prefix. For consistency with `PUSH_ENDPOINT` that already exists I suggest the `PUSH_` prefix.

One question: is the agent able to work with `poll` mode as well? ie if it's configured to use `poll` instead of `push`, will it query a new overlay every minute and push the result to its connected cloudservers?

jeremyds commented:
@RachedBenMustapha:
> One question: is the agent able to work with poll mode as well? ie if it's configured to use poll instead of push, will it query a new overlay every minute and push the result to its connected cloudservers?

No. I did not handle the poll mode, I thought it was "unused" (the previous mode). I don't see a big work to make it works, I just need to call the agent listener callback.

rachedbenmustapha commented:
@jeremyds ok, if we switch off the agent-less management mode then the agent should pick up the duty of initiating poll mode (like startManagementListeners does).

It's not enabled by default and event discouraged, but it's still important to support as the kube deployment has configuration specific to this mode.

jeremyds commented:
@RachedBenMustapha: 

> For consistency with PUSH_ENDPOINT that already exists I suggest the PUSH_ prefix.

I don't get this one :S, which env variable are you speaking about? maybe ORBIT_PUSH_ENDPOINT or
 .MANAGEMENT_ENDPOINT?

rachedbenmustapha commented:
it's just called `PUSH_ENDPOINT` and is in `lib/management/index.js` (maybe it has disappeared in 8.1? if that's the case it needs to come back)

jeremyds commented:
> if we switch off the agent-less management mode then the agent should pick up the duty of initiating poll mode (like startManagementListeners does).

It's not enabled by default and event discouraged, but it's still important to support as the kube deployment has configuration specific to this mode.


The current code initailize the POLL management when the appropriate env variable is set. But in that case the new overlays received in poll mode don't trigger the agent listener callback and so, are not broadcasted to agent clients (however they are saved..... and patched). I did not handle at all the POLL mode

rachedbenmustapha commented:
Just verified about `PUSH_ENDPOINT`, this PR's commit "ft: ZENKO-716: management, use constant and env variables for tests" removed it, it needs to come back to keep compatibility and not upset the whole team who are using it 😆

jeremyds commented on lib/management/index.js:
```diff
@@ -13,19 +13,16 @@ const { initManagementCredentials } = require('./credentials');
+const managementEndpoint = `${managementEndpointRoot}/${endpointPath}`;
 
-const pushEndpointRoot =
-    process.env.PUSH_ENDPOINT ||
```
@RachedBenMustapha here is the culprit

jeremyds replied:
you we're faster to find it :(

jeremyds commented:
> Just verified about `PUSH_ENDPOINT`, this PR's commit "ft: ZENKO-716: management, use constant and env variables for tests" removed it, it needs to come back to keep compatibility and not upset the whole team who are using it

It is to be sure they update their working version 8-)

jeremyds committed:
ft: ZENKO-716: rename env variablle for coherence.

jeremyds committed:
ft: ZENKO-716: new management process poll mode

This commit fixes the management pull mode of the new management
process. We don't need anymore to save and patch the overlay into
lib/management/poll.

Furthermore, like for the push mode, a callback should be fired when a
new overlay is received, to notify the upper layer. The initManagement
takes this new callback as parameter and the previous listener
registering in lib/management/push is no longer required.

rachedbenmustapha commented on lib/management/push.js:
```diff
@@ -86,13 +84,17 @@ function startWSManagementClient(pushEndpoint, instanceId, token) {
     const agent = createWSAgent(pushEndpoint, process.env, logger);
 
-    const url = `${pushEndpoint}/${instanceId}/ws`;
+    const url = process.env.ORBIT_PUSH_ENDPOINT ||
```
The env var that already exists and is used by kube deployment is called `PUSH_ENDPOINT`

jeremyds replied:
Done

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rahulreddy

* @dora-korpar

* @RachedBenMustapha








*The following options are set:* **after_pull_request**

rachedbenmustapha commented on lib/management/index.js:
```diff
@@ -13,19 +13,16 @@ const { initManagementCredentials } = require('./credentials');
-const managementEndpoint = `${managementEndpointRoot}/api/v1/instance`;
+const managementEndpoint = `${managementEndpointRoot}/${endpointPath}`;
 
-const pushEndpointRoot =
```
I did not see this when reviewing just the last 2 commits: I meant to keep this here as `PUSH_ENDPOINT` instead of moving to a `ORBIT_PUSH_ENDPOINT` lookup in `startWSManagementClient` and hardcoding this one here.

So to be clear the code here should read:
```js
const pushEndpointRoot = process.env.PUSH_ENDPOINT || 'https://push.api.zenko.io';
const pushEndpoint = `${pushEndpointRoot}/api/v1/instance`;
```

jeremyds replied:
done

rachedbenmustapha commented on lib/management/push.js:
```diff
@@ -86,13 +84,17 @@ function startWSManagementClient(pushEndpoint, instanceId, token) {
     const agent = createWSAgent(pushEndpoint, process.env, logger);
 
-    const url = `${pushEndpoint}/${instanceId}/ws`;
+    const url = process.env.PUSH_ENDPOINT ||
```
`${pushEndpoint}/${instanceId}/ws` should be used here as it has been correctly computed, and connecting the websocket client to `process.env.PUSH_ENDPOINT` directly will not work (it's missing paths).

jeremyds replied:
done

jeremyds committed:
ft: ZENKO-716: endpoint value

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rahulreddy

* @RachedBenMustapha

* @dora-korpar








*The following options are set:* **after_pull_request**

jeremyds committed:
ft: ZENKO-716: fix setTimeout

jeremyds committed:
ft: ZENKO-716: fix infinite loop in initManagement.

The initManagement function has an automatic retry when it fails. The
problem is that when an INITIAL_INSTANCE_ID environment variable is
passed and there is already an instance ID saved in metadata and these
two values are different, the initManagement will retry infinitely.

This commit catches this situation and make the initManagement function
stop to retry and instead it raised an error to upper level. This error
is now caught in upper level and leads to a process exit.

jeremyds committed:
ft: ZENKO-716: management process, handle browserAccess change

The overlay sent by orbit to Zenko contains a "browserAccess" field.
When this parameter is set to false, payload message sent by Orbit
should not be forwarded. With the code move from the cloudserver to the
new management process, this value should be handled and the appropriate
configuration event should be emitted on value change.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @dora-korpar

* @RachedBenMustapha

* @rahulreddy








*The following options are set:* **after_pull_request**

jeremyds commented on lib/management/index.js:
```diff
@@ -66,36 +57,39 @@ function startManagementListeners(instanceId, token) {
+        },
+        (instanceId, cb) => initManagementCredentials(instanceId, log, cb),
         (instanceId, token, cb) => {
             if (!isManagementAgentUsed()) {
```
should be removed, error while resolving conflict after rebasing on 8.1
